### PR TITLE
Upmerge fix

### LINKED ISSUE #1632: Fix upmerge samples to edge workflow
### Steps to reproduce (required)
The upmerge samples to edge workflow fails when there's nothing to commit. This is okay since there were technically no changes but the workflow should still pass in this case 

https://github.com/radius-project/samples/actions/workflows/upmerge.yaml

## Additional context

[AB#12892](https://dev.azure.com/azure-octo/e61041b4-555f-47ae-95b2-4f8ab480ea57/_workitems/edit/12892)

### PR DESCRIPTION
**Pull request by brooke-hamilton**
# Description

Fixes an issue in which the [Upmerge samples to edge action](https://github.com/radius-project/samples/actions/workflows/upmerge.yaml) completes with an error when there are no changes to merge.

## Type of change

- This pull request fixes a bug in Radius and has an approved issue (issue link required).

Fixes: #1632 

## Changes

- The `upmerge.yaml` file has been updated to check for changes before attempting to create a new PR.
- Comments added to explain context and actions taken in the workflow.
- Merge and commit in one `git` command instead of two. Not all merges create a merge commit, so this allows for specifying a merge comment that will only be used if a commit is created by the merge.
- Other minor code improvements.

## Background context

The [Upmerge samples to edge action](https://github.com/radius-project/samples/actions/workflows/upmerge.yaml) workflow is part of the release process. The release engineer uses this workflow to synchronize any changes made to the default branch to the `edge` branch. That step is necessary because the default branch is also the current release branch, which allows PRs to be approved against the current release and the samples to be immediately updated (rather than having to wait until the next release for a fix/update to the samples.) After synchronizing changes to the `edge` branch, the next release branch is created from the edge branch.

## Other considerations

There is potential for confusion for the person executing the workflow if they run the automation, but there are no changes between the source branch and the target branch, and the workflow completes successfully without creating a PR. The user would have to look at the workflow log to see that no PR was created because there are no changes. To make that experience better I tested a suggestion from @ytimocin to force the PR to be created without any files to merge. However, GitHub will not allow an empty PR to be created even while passing a `-f` flag to the `gh pr create` command.

@ytimocin, @willdavsmith thank you for early feedback on this PR. 🚀

### CODE REVIEW

File: .github/workflows/upmerge.yaml
Comment (ytimocin) [inline .github/workflows/upmerge.yaml:12]: Should these be added to the secrets of the repo? Would that create a problem in the forks if we do that?

File: .github/workflows/upmerge.yaml
Comment (ytimocin) [inline .github/workflows/upmerge.yaml]: Is the source branch **edge** or the **release branch**? Because if it is **edge** and if we create a branch from **edge**, the new branch will not have any changes.

The documentation is great by the way. Thanks! Can we also add an example like main branch is 0.36, edge, new branch...

File: .github/workflows/upmerge.yaml
Comment (ytimocin) [inline .github/workflows/upmerge.yaml]: `github.ref` branch being release branch?

File: .github/workflows/upmerge.yaml
Comment (ytimocin) [inline .github/workflows/upmerge.yaml:73]: 👏

**Review by ytimocin — COMMENTED** [review]
Looking great so far!

File: .github/workflows/upmerge.yaml
Comment (brooke-hamilton) [reply .github/workflows/upmerge.yaml]: ✅ I added an example and clarification about which branch is the release branch.

File: .github/workflows/upmerge.yaml
Comment (brooke-hamilton) [reply .github/workflows/upmerge.yaml]: I clarified this language. Thank you! 🎆

File: .github/workflows/upmerge.yaml
Comment (brooke-hamilton) [reply .github/workflows/upmerge.yaml:73]: GitHub Copilot may have had a hand in this. 🥇

File: .github/workflows/upmerge.yaml
Comment (brooke-hamilton) [reply .github/workflows/upmerge.yaml:12]: I think this is a very good idea. How about this: at the top of all the workflows we declare which environment variables are required to run the workflow. When someone forks the repo, they would go to the forked repo (in GitHub) --> settings --> Secrets and Variables --> Actions --> Variables and add the variables needed, like username and email. (Secrets can be configured the same way.)

We would have to coordinate the creation of these secrets in the origin repo before merging a change like that.

Comment (ytimocin) [reply .github/workflows/upmerge.yaml:12]: You can create an issue in the samples repo to track this work if you would like. You can also post the issue link here.

File: .github/workflows/upmerge.yaml
Comment (ytimocin) [inline .github/workflows/upmerge.yaml:6]: So, this is done right after we do the release, right? Let's say we have **_just_** finished releasing 0.36 and 0.36 becomes the new main branch. We kick this workflow off from the 0.36 branch, right? All the commits from edge gets merged into the new release branch i.e. 0.36 in this instance.

- 0.35 was the main branch
- edge has 6 commits on top of 0.35
- we release 0.36
- now 0.36 branch is the main branch
- this workflow is kicked off from 0.36
- 0.36 gets rebased with the branch created by 0.36

Is this the flow?

Comment (brooke-hamilton) [reply .github/workflows/upmerge.yaml:6]: We do not have a formal document on the release process in this repo, but there is a similar process [documented in the docs repo](https://github.com/radius-project/docs/blob/v0.36/docs/content/contributing/docs/maintainers-docs/index.md). In that process, the steps are like this:

Regular PR process:
- v0.35 is the default (main) branch.
- The edge branch was equivalent to v0.35 when the release was created.
- Changes are PR'd to the v0.35 branch and the edge branch falls behind.

Process for creating a new release:
- Create a new branch from edge.
- Merge changes from v0.35 to the branch just created from edge.
- PR the changes to edge from the branch just created. (When this is complete, edge will be at the same commit as v0.35.)
- Create the new release branch, v0.36, as a branch from edge.
- Update the v0.36 branch to be the default (main) branch.

At that point in time, v0.35, v0.36, and edge will all be at the same commit. v0.36 will be the new default branch. New PRs will go to the v0.36 branch.

The `upmerge.yaml` workflow was created to support the process described above. @willdavsmith please correct anything written above.

```mermaid
%%{init: { 'logLevel': 'debug', 'theme': 'base', 'gitGraph': {'showBranches': true, 'showCommitLabel':true,'mainBranchName': 'edge'}} }%%
gitGraph LR:
    commit id: "Release"
    branch v0.35
    commit id: "PR 1"
    commit id: "PR 2"
    commit id: "PR 3"
    checkout edge
    branch edge-next1
    merge v0.35
    checkout edge
    merge edge-next1
    branch v0.36
    commit id: "PR 4"
    commit id: "PR 5"
    commit id: "PR 6"
    checkout edge
    branch edge-next2
    merge v0.36
    checkout edge
    merge edge-next2
    branch v0.37
    commit id: "PR 7"
    commit id: "PR 8"
    commit id: "PR 9"
```

Comment (ytimocin) [reply .github/workflows/upmerge.yaml:6]: What I was thinking is that once a release branch is created, it shouldn't receive any more commits because that branch is like what we have for that specific release.

Then **edge** starts getting the new commits which will eventually become the new release.

Example:

1. 0.35 (main branch) - **30 commits**.
2. All the new PRs get based off of **edge**.
3. Let's assume just before the 0.36 release, edge has 5 more commits on top of the main branch (0.35 branch) meaning **35 commits**.
4. We do the 0.36 release.
5. 0.36 release (main branch) - **35 commits**.
6. Go back to item 2 all the way to item 4.

Does that make sense? cc/ @willdavsmith @willtsai @rynowak

Comment (ytimocin) [reply .github/workflows/upmerge.yaml:6]: > At that point in time, v0.35, v0.36, and edge will all be at the same commit. v0.36 will be the new default branch. New PRs will go to the v0.36 branch.

I don't see a reason why all 3 branches need to be at the same commit. 0.35 should be behind and edge and 0.36 should be at the same commit at the time of 0.36 release.

File: .github/workflows/upmerge.yaml
Comment (willdavsmith) [inline .github/workflows/upmerge.yaml:77]: we should also update the release process doc to inform that there might not be a PR created: https://github.com/radius-project/radius/tree/main/docs/contributing/contributing-releases

**Review by willdavsmith — APPROVED** [review]
LGTM, awesome stuff

**Pull request merged**
